portal-impl/src/com/liferay/portal/layoutconfiguration/util/PortletRenderer.java

```diff
@@ -108,6 +115,46 @@ public StringBundler renderError(
 		}
 	}
 
+	public Map<String, Object> renderHeaders(
+			HttpServletRequest request, HttpServletResponse response)
+		throws PortletContainerException {
+
+		request = PortletContainerUtil.setupOptionalRenderParameters(
+			request, null, _columnId, _columnPos, _columnCount);
+
+		BufferCacheServletResponse bufferCacheServletResponse =
+			new BufferCacheServletResponse(response);
+
+		PortletContainerUtil.renderHeaders(
+			request, bufferCacheServletResponse, _portlet);
+
+		Map<String, Object> headerRequestAttributes = new HashMap<>();
+
+		Enumeration<String> attributeNames = request.getAttributeNames();
```

Comment:
@ngriffin7a this seems a bit weird to me, do we really need to copy all the attributes from the request? This is going to be super heavy.

Reply:
@shuyangzhou This is necessary because the JSR 378 JSF Portlet Bridge has to run in the `HEADER_PHASE` of the portlet lifecycle in order to ensure that JS/CSS resources are added via `HeaderResponse.addDependency()`. It saves the rendered markup in a request attribute that is later rendered to the response in the `RENDER_PHASE` of the portlet lifecycle.

Reply:
@ngriffin7a I know there must be something in the attributes that we need to carry on, but blindly copying over all the attributes is a very heavy solution. Anyway we can only copy over what is known to be needed?

Reply:
@shuyangzhou See https://github.com/liferay/liferay-faces-bridge-impl/blob/master/bridge-impl/src/main/java/com/liferay/faces/bridge/internal/BridgePhaseHeaderRenderCommon.java#L282 but don't let the `renderRequest` variable name confuse you. It is a `HeaderRequest` that is being referenced. `HeaderRequest` extends `RenderRequest` so I used `renderRequest` as a variable name.